File: matlab/ode45/answer_entrain.m
function answer_entrain
    filename='littlerock.nc';
    fprintf('reading file: %s\n',filename);
    file_struct=nc_info(filename);
    c=constants;
    %
    % grab the March 2 12Z sounding
    %
    sound_var = file_struct.Dataset(4).Name;
    fprintf('found sounding: %s\n',sound_var);
    press=nc_varget(filename,sound_var,[0,0],[Inf,1]);
    height=nc_varget(filename,sound_var,[0,1],[Inf,1]);
    temp=nc_varget(filename,sound_var,[0,2],[Inf,1]);
    dewpoint=nc_varget(filename,sound_var,[0,3],[Inf,1]);
    envHeight=nudgeHeight(height);
    interpTenv=@(zVals) interp1(envHeight,temp,zVals);
    interpTdEnv=@(zVals) interp1(envHeight,dewpoint,zVals);
    interpPress=@(zVals) interp1(envHeight,press,zVals);
    p900_level=find(abs(900 - press) < 2.);
    p800_level=find(abs(800 - press) < 7.);
    thetaeVal=thetaep(dewpoint(p900_level) + c.Tc,temp(p900_level) + c.Tc,press(p900_level)*100.);
    height_800=height(p800_level);
    wTcloud=wsat(dewpoint(p900_level) + c.Tc,press(p900_level)*100.);
    yinit=[0.5,height_800,thetaeVal,wTcloud];
    tspan=0:10:2500;
    lambda=2.e-4;
    derivs=@(t,y) F(t,y,lambda,interpTenv,interpTdEnv,interpPress);
    options=odeset('Events',@stopIt);
    [t,y]=ode45(derivs,tspan,yinit,options);
    wvel=y(:,1);
    cloud_height=y(:,2); 
    thetae_cloud=y(:,3);
    wT_cloud=y(:,4);

    %limit the plot to the updraft
    figure(1);
    clf;
    plot(wvel,cloud_height)
    xlabel('vertical velocity');
    ylabel('height above surface');
    title('wvel vs. height')
    grid on;
    
    for i=1:numel(cloud_height)
      the_press=interpPress(cloud_height(i))*100.;
      [Tcloud(i),wvCloud(i),wlCloud(i)]= ...
         tinvert_thetae(thetae_cloud(i),wT_cloud(i),the_press);
    end
    for i=1:numel(cloud_height)
      the_press=interpPress(cloud_height(i))*100.;
      [Tadia(i),wvAdia(i),wlAdia(i)]= ...
         tinvert_thetae(thetae_cloud(1),wT_cloud(1),the_press);
    end

    figure(2);
    clf;
    plot(Tcloud - c.Tc,cloud_height,'r-')
    hold on;
    plot(temp,envHeight,'g-');
    plot(Tadia - c.Tc,cloud_height,'b-');
    xlabel('vertical velocity');
    ylabel('height above surface');
    title('wvel vs. height')
    grid on;
    hold off;
    
    function [value,isterminal,direction] = stopIt(t,y)
        % Locate the time when height passes through stopHeight in a decreasing direction
        % and stop integration.  Here we use a nested function to avoid
        % passing the additional parameter stopHeight as an input argument.
        value = y(1);     % detect height = 0
        isterminal = 1;   % stop the integration
        direction = -1;   % only when approaching stopHeight from above
   end

end

function yp=F(t,y,lambda,interpTenv,interpTdEnv,interpPress)
  c=constants;
  yp=zeros(4,1); % since output must be a column vector
  %first variable is velocity in m/s
  height=y(2);
  thetae_cloud=y(3);
  wT_cloud=y(4);
  yp(1)=calcBuoy(height,thetae_cloud,wT_cloud,interpTenv,interpTdEnv,interpPress);
  press=interpPress(height)*100.;%hPa
  Tdenv=interpTdEnv(height) + c.Tc;
  Tenv=interpTenv(height) + c.Tc;
  wTenv=wsat(Tdenv,press); %kg/kg
  thetaeEnv=thetaep(Tdenv,Tenv,press);
  %second variable is height in m
  yp(2)= y(1);
  yp(3)=lambda*(thetaeEnv - y(3));
  yp(4)= lambda*(wTenv - y(4));
end

function Bout=calcBuoy(height,thetae0,wT0,interpTenv,interpTdEnv,interpPress)
    %calcBuoy(height,thetae0,wT0,interpTenv,interpTdenv,interpPress)
    %input: height (m), thetae0 (K), plus function handles for
    %T,Td, press soundings
    %output: Bout = buoyant acceleration in m/s^2
    %neglect liquid water loading in the virtual temperature
    c=constants;
    press=interpPress(height)*100.;%hPa
    [Tcloud,wvCloud,wlCloud]=tinvert_thetae(thetae0,wT0,press); %K
    %wvcloud=wsat(Tcloud,press); %kg/kg
    Tvcloud=Tcloud*(1. + c.eps*wvCloud - wlCloud);
    Tenv=interpTenv(height) + c.Tc;
    Tdenv=interpTdEnv(height) + c.Tc;
    wvenv=wsat(Tdenv,press); %kg/kg
    Tvenv=Tenv*(1. + c.eps*wvenv);
    TvDiff=Tvcloud - Tvenv;
    fprintf('%10.3f %10.3f %10.3f\n',press*0.01,height,TvDiff);
    Bout=c.g0*(TvDiff/Tvenv);
end


function envHeight=nudgeHeight(zVec)
    %if two balloon pressure levels are idential
    %add a factor of 0.1% to the second one
    %so interpolation will work
    envHeight=zVec;
    hit=find(diff(envHeight) < 0);
    envHeight(hit+1)=zVec(hit) + 1.e-3*zVec(hit);
end

File: python/ode45/answer_entrain.py
from netCDF4 import Dataset
import site
site.addsitedir('C:\Users\Den\mya405\python\\thermlib')
site.addsitedir('C:\Users\Den\mya405\python\\skew_T')
from scipy.integrate import ode
import matplotlib.pyplot as plt
import numpy as np
from constants import constants as c
from nudge import nudge
from new_thermo import thetaep, tinvert_thetae, wsat
from calcBuoy import calcBuoy
from findTmoist import findTmoist

def answer_entrain():
    filename = 'littlerock.nc'
    print 'reading file: %s\n' %(filename)
    nc_file = Dataset(filename)
    var_names = nc_file.variables.keys()
    print nc_file.ncattrs()
    print nc_file.units
    print nc_file.col_names
    
    sound_var = nc_file.variables[var_names[3]]
    press = sound_var[:,0]
    height = sound_var[:,1]
    temp = sound_var[:,2]
    dewpoint = sound_var[:,3]
    
    #height must have unique values
    envHeight= nudge(height)
    #Tenv and TdEnv interpolators return temp. in deg C, given height in m
    #Press interpolator returns pressure in hPa given height in m
    interpTenv = lambda zVals: np.interp(zVals, envHeight, temp)
    interpTdEnv = lambda zVals: np.interp(zVals, envHeight, dewpoint)
    interpPress = lambda zVals: np.interp(zVals, envHeight, press)
    
    p900_level = np.where(abs(900 - press) < 2.)
    p800_level = np.where(abs(800 - press) < 7.)
    thetaeVal=thetaep(dewpoint[p900_level] + c.Tc,temp[p900_level] + c.Tc,press[p900_level]*100.)
    height_800=height[p800_level]
    wTcloud = wsat(dewpoint[p900_level] + c.Tc, press[p900_level]*100.)
    entrain_rate = 2.e-4
    winit = 0.5 #initial velocity (m/s)
    yinit = [winit, height_800, thetaeVal, wTcloud]  
    tinit = 0
    tfin = 2500
    dt = 10
    
    #want to integrate F using ode45 (from MATLAB) equivalent integrator
    r = ode(F).set_integrator('dopri5')
    r.set_f_params(entrain_rate, interpTenv, interpTdEnv, interpPress)
    r.set_initial_value(yinit, tinit)
    
    y = np.array(yinit)
    t = np.array(tinit)
    
    #stop tracking the parcel when the time runs out, or if the parcel stops moving/is desecnding
    while r.successful() and r.t < tfin and r.y[0] > 0:
        #find y at the next time step
        #(r.integrate(t) updates the fields r.y and r.t so that r.y = F(t) and r.t = t 
        #where F is the function being integrated)
        r.integrate(r.t+dt)
        if r.y[0] <= 0:
            break
        #keep track of y at each time step
        y = np.vstack((y, r.y))
        t = np.vstack((t, r.t))
        
    wvel = y[:,0]
    cloud_height = y[:,1]
    thetae_cloud = y[:,2]
    wT_cloud = y[:,3]
    
    plt.figure(1)
    plt.plot(wvel, cloud_height)
    plt.xlabel('vertical velocity (m/s)')
    plt.ylabel('height above surface (m)')
    plt.gca().set_title('vertical velocity of a cloud parcel vs height,\
 entrainment rate of %4.1e $s^{-1}$' %entrain_rate)
    
    Tcloud = np.zeros(cloud_height.size)
    wvCloud = np.zeros(cloud_height.size)
    wlCloud = np.zeros(cloud_height.size)
    
    for i in range(0, len(cloud_height)):
        the_press = interpPress(cloud_height[i])*100.
        Tcloud[i], wvCloud[i], wlCloud[i] = tinvert_thetae(thetae_cloud[i], 
                                                        wT_cloud[i], the_press)
        
    Tadia= np.zeros(cloud_height.size)
    wvAdia = np.zeros(cloud_height.size)
    wlAdia = np.zeros(cloud_height.size)
    
   
    for i in range(0, len(cloud_height)):
        the_press = interpPress(cloud_height[i])*100.
        Tadia[i], wvAdia[i], wlAdia[i] = tinvert_thetae(thetae_cloud[0], 
                                                      wT_cloud[0], the_press)
    
    plt.figure(2)
    TcloudHandle, = plt.plot(Tcloud - c.Tc, cloud_height, 'r-')
    TenvHandle, = plt.plot(temp, envHeight, 'g-')
    TadiaHandle, = plt.plot(Tadia - c.Tc, cloud_height, 'b-')
    plt.xlabel('temperature (deg C)')
    plt.ylabel('height above surface (m)')
    plt.gca().set_title('temp. of rising cloud parcel vs height,\
 entrainment rate of %4.1e $s^{-1}$' %entrain_rate)
    plt.gca().legend([TcloudHandle, TenvHandle, TadiaHandle],['cloud', 'environment', 'moist adiabat'])
    plt.show()
    
       
#F returns the buoyancy (and height), velocity, rate of change of thetae_cloud 
#(w.r.t. time) and rate of change of total mixing ratio at a given time step and height
def F(t, y, entrain_rate, interpTenv, interpTdEnv, interpPress):
    yp = np.zeros((4,1))
    velocity = y[0]
    height = y[1]
    thetae_cloud = y[2]
    wT_cloud = y[3]
    #yp[0] is the acceleration, in this case the buoyancy 
    yp[0] = calcBuoy(height, thetae_cloud, interpTenv, interpTdEnv, interpPress)
    press = interpPress(height)*100. #Pa
    Tdenv = interpTdEnv(height) + c.Tc #K
    Tenv = interpTenv(height) + c.Tc #K
    wTenv = wsat(Tdenv, press) #kg/kg
    thetaeEnv = thetaep(Tdenv, Tenv, press)
    #yp[1] is the rate of change of height
    yp[1] = velocity
    #yp[2] is the rate of change of thetae_cloud
    yp[2] = entrain_rate*(thetaeEnv - thetae_cloud)
    #yp[3] is the rate of change of wT_cloud
    yp[3] = entrain_rate*(wTenv - wT_cloud)
    return yp

if __name__ == "__main__":
    answer_entrain()







